src/Data/Functor/Foldable.hs

```diff
@@ -181,29 +301,77 @@ class Functor (Base t) => Recursive t where
     -> a
   gprepro k e f = extract . c where c = fmap f . k . fmap (duplicate . c . hoist e) . project
 
+-- | TODO
```

Comment:
I don't understand what all the `dist` stuff is about. Could someone enlighten me?

Reply:
The README provides a quick intro:

> One way to implement such a custom recursion scheme is to combine the features of existing recursion schemes. For example, a "paramorphism" gives the non-recursive function access to the original sub-trees, a "zygomorphism" gives that function access to auxiliary results computed from those sub-trees, and so the combined "zygomorphic paramorphism" gives that function access to both the original sub-trees and the auxiliary results. In order to construct such combinations, most of our recursion schemes come in a generalized variant, e.g. gzygo, and in a "distributive law transformer" variant, e.g. distZygoT. Just like monad transformers, distributive law transformers can be combined into stacks, and like monad transformers, the order in which you combine the layers determines how the layers interact with each other. Apply a generalized recursion scheme to a stack of distributive laws in order to obtain a recursion scheme which has both the features of the generalized recursion scheme and those of the distributive laws.

It's based on a paper called "recursion schemes from comonads".

Reply:
The idea is that a comonad is _almost_ enough to define a recursion scheme, and although they don't say this in the paper, that's appealing because we have comonad transformers, so maybe we can combine comonads in order to combine recursion schemes.

Reply:
But what's the relationship between comonads and recursion schemes? Well, consider `zygo`: compared to `cata`, we're carrying around an extra `(b,)`. That's a comonad. And consider `histo`: compared to `cata`, we're carrying around an extra `Cofree (Base t)`; that's another comonad.

Reply:
So the idea behind `gcata` is that it generalizes `cata`, `zygo`, `histo`, and so on, and we can choose which one by instantiating its `w` comonad to `Identity`, `(b,)`, or `Cofree (Base t)`. We can also instantiate it to other comonads, or to stacks of comonad transformers, in order to obtain novel recursion schemes.

Except it turns out a comonad is not enough! We need a _distributive_ comonad. That's where the distributive laws come in.